content/courses/advanced-ada/parts/resource_management/access_types.rst

```diff
@@ -3445,26 +3445,239 @@ caller.
+general, however, we should avoid this approach. Instead, our recommendation is
+to encapsulate the access types in some form of abstraction. In this section,
+we discuss design strategies for access types that take this recommendation
+into account.
+
+Abstract data type for access types
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The simplest form of abstraction is of course an abstract data type. For
+example, we could declare a limited private type, which allows us to hide
+the access type and to avoid copies of references that could potentially
+become dangling references. (We discuss limited private types later
+:ref:`in another chapter <Adv_Ada_Limited_Private_Types>`.)
+
+Let's see an example:
+
+.. code:: ada run_button project=Courses.Advanced_Ada.Resource_Management.Access_Types.Design_Strategies.Access_Type_Abstraction
+
+    package Access_Type_Abstraction is
+
+       type Info is limited private;
+
+       function Init (I : Integer) return Info;
+
+       function Copy (Obj : Info) return Info;
+
+       procedure Destroy (Obj : in out Info);
+
+       function To_Integer (Obj : Info)
+                            return Integer;
+
+    private
+
+       type Info is access Integer;
+
+    end Access_Type_Abstraction;
+
+    with Ada.Unchecked_Deallocation;
+
+    package body Access_Type_Abstraction is
+
+       function Init (I : Integer) return Info is
+       begin
+          return new Integer'(I);
+       end Init;
+
+       function Copy (Obj : Info) return Info is
+       begin
+           return Init (Obj.all);
+       end Copy;
+
+       procedure Destroy (Obj : in out Info) is
+          procedure Free is
+            new Ada.Unchecked_Deallocation
+              (Object => Integer,
+               Name   => Info);
+       begin
+          Free (Obj);
+       end Destroy;
+
+       function To_Integer (Obj : Info)
+                            return Integer is
+       begin
+          return Obj.all;
+       end To_Integer;
+
+    end Access_Type_Abstraction;
+
+    with Ada.Text_IO; use Ada.Text_IO;
+
+    with Access_Type_Abstraction;
+    use  Access_Type_Abstraction;
+
+    procedure Show_Access_Type_Abstraction is
+       Obj_1 : Info := Init (100);
+       Obj_2 : Info := Copy (Obj_1);
+    begin
+       Put_Line ("Obj_1 : " &
+                 To_Integer (Obj_1)'Image);
+       Put_Line ("Obj_2 : " &
+                 To_Integer (Obj_2)'Image);
+       Destroy (Obj_1);
+       Destroy (Obj_2);
+    end Show_Access_Type_Abstraction;
+
+In this example, we hide an access type in the :ada:`Info` type |mdash| a
+limited private type. We allocate an object of this type in the :ada:`Init`
+function and deallocate it in the :ada:`Destroy` procedure. Also, we make
+sure that the reference isn't copied in the :ada:`Copy` function |mdash|
+we only copy the designated value in this function. This strategy eliminates
+the possibility of dangling references, as each reference is encapsulated in
```

Comment:
I don't understand how that makes dangling references impossible. For example, given an object named X, of type Info, wouldn't a call to Destroy(X), followed by a call to To_Integer(X), dereference a null access value, ie a dangling reference?

Reply:
Yes, you're right: this code example was bad. I've just reworked it and improved it a bit.